doc/hardware/peripherals/sensor/read_and_decode.rst

```diff
@@ -0,0 +1,208 @@
+
+    #include <zephyr/drivers/sensor.h>
+
+    #define TEMP_CHANNEL { SENSOR_CHAN_AMBIENT_TEMP, 0 }
+
+    SENSOR_DT_READ_IODEV(dht_iodev, DT_ALIAS(dht0), {TEMP_CHANNEL});
+    RTIO_DEFINE(dht_ctx, 1, 1);
+
+    int main (void)
+    {
+        int rc;
+        uint8_t buf[8];
+        uint32_t temp_frame_iter = 0;
+        struct sensor_q31_data temp_data = {0};
+        struct sensor_decoder_context temp_decoder = SENSOR_DECODER_CONTEXT_INIT(
+            SENSOR_DECODER_DT_GET(DT_ALIAS(dht0)),
+            buf,
+            SENSOR_CHAN_AMBIENT_TEMP,
+            0
+        );
+
+
+        while (1) {
+            /* Blocking read */
+            rc = sensor_read(dht_iodev, &dht_ctx, buf, sizeof(buf));
+
+            if (rc != 0) {
+                printk("sensor_read() failed %d\n", rc);
+            }
+
+            /* Decode the data into a single q31 */
+            sensor_decode(&temp_decoder, &temp_data, 1);
+
+            printk("Temperature " PRIsensor_q31_data "\n" ,
+                   PRIsensor_q31_data_arg(temp_data, 0));
+
+            k_msleep(1);
+        }
+
+    }
+
+Polling Read with Multiple Sensors
+**********************************
+
+One of the benefits of Read and Decode is the ability to concurrently read many
+sensors in one thread. Effectively read requests are started asynchronously for
+all sensors. When each read completes we then decode the sensor data. Examples
+speak loudly and so a sample showing how this might work:
+
+.. code-block:: c
+
+    #include <zephyr/drivers/sensor.h>
+
+    #define TEMP_CHANNEL { SENSOR_CHAN_AMBIENT_TEMP, 0 }
+
+    #define DHT_ALIAS(id) DT_ALIAS(CONCAT(dht, id))
+    #define DHT_IODEV_SYM(id) CONCAT(dht_iodev, id)
+    #define DHT_IODEV_PTR(id) &DHT_IODEV_SYM(id)
+    #define DHT_DEFINE_IODEV(id) \
+        SENSOR_DT_READ_IODEV( \
+            DHT_IODEV_SYM(id), \
+            DHT_ALIAS(id), \
+            {TEMP_CHANNEL} \
+        );
+
+    #define NUM_SENSORS 8
+
+    LISTIFY(NUM_SENSORS, DHT_DEFINE_IODEV, (;));
+
+    struct sensor_iodev *iodevs[NUM_SENSORS] = {LISTIFY(NUM_SENSORS, DHT_IODEV_PTR, (,)};
+
+    RTIO_DEFINE_WITH_MEMPOOL(dht_ctx, NUM_SENSORS, NUM_SENSORS, NUM_SENSORS, 8, sizeof(void*));
+
+    int main (void)
+    {
+        int rc;
+        uint32_t temp_frame_iter = 0;
+        struct sensor_q31_data temp_data = {0};
+        struct sensor_decoder_api *decoder;
+        struct rtio_cqe *cqe;
+        uint8_t *buf;
+        uint32_t buf_len;
+
+        while (1) {
+            /* Non-Blocking read for each sensor */
+            for (int i = 0; i < NUM_SENSORS; i++) {
+                rc = sensor_read_async_mempool(iodevs[i], &dht_ctx, iodevs[i]);
+
+                if (rc != 0) {
+                    printk("sensor_read() failed %d\n", rc);
+                    return;
+                }
+            }
+
+            /* Wait for read completions */
+            for (int i = 0; i < NUM_SENSORS; i++) {
+               cqe = rtio_cqe_consume_block(&dht_ctx);
```

Comment:
Should we encourage users to rather utilize the wrapped version of this API: `sensor_processing_callback()`?

Reply:
I've been mulling over the callback, I'm not a huge fan as it sort of disrupts the linear code flow and calls off to somewhere else. Have long thought what would be nice is more of an endless iterator/generator style loop akin to ...

```c
sensor_consume_mempool_read(rc, buf, buf_len, userdata) {
   /* my logic here */
}
```

Reply:
Yeah - I think we should document it somewhere on this guide, since it's what we have as of today. Perhaps an alternative method living alongside the `rtio_cqe_consume_block()`?